File: Exercise_2.cpp
/* 
    Time Complexity                              :  O(2*N), in worst case it will iteratve over the array twice if the biggest element is the last element.
    Space Complexity                             :  O(N) store all the n-1 values in the stack if the nth value is the largest and all the other elements are in desecending or equal.
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/next-greater-element-ii/

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n,-1);
        stack<int> st;
        int i=0;
        while(i < 2*n) {
            int cur = nums[i%n];
            while(!st.empty() and nums[st.top()] < cur) {
                if(ans[st.top()] == -1) {
                    ans[st.top()] = cur;
                }
                st.pop();
            }
            if(!st.empty() and i%n == st.top()) {
                break;
            }
            if(ans[i%n] == -1)
                st.push(i%n);
            i++;
        }
        
        return ans;
    }
};

File: Exercise_2.py
# Problem2 Next Greater Element II (https://leetcode.com/problems/next-greater-element-ii/)
# // Time Complexity :O(2N) = O(n)
# // Space Complexity :O(n)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No
class Solution:
    def nextGreaterElements(self, nums):
        n = len(nums)
        # using stack
        stack = []
        # creating result list
        result = [-1] * n
        # iterating over the array twice coz rotaional
        for i in range(2 * n):
            # till the curr temp is greater than last element in stack keep popping
            while stack and nums[stack[-1]] < nums[i % n]:
                result[stack.pop()] = nums[i % n]
            if i < n:
                stack.append(i)
        return result


sol = Solution()
print(sol.nextGreaterElements([1, 2, 1]))
